enterprise/frontend/src/embedding-sdk/components/public/QueryVisualization.tsx

```diff
@@ -0,0 +1,165 @@
+import { useEffect, useState } from "react";
+
+import LoadingAndErrorWrapper from "metabase/components/LoadingAndErrorWrapper";
+import { useSelector } from "metabase/lib/redux";
+import {
+  onCloseChartType,
+  onOpenChartSettings,
+  setUIControls,
+} from "metabase/query_builder/actions";
+import QueryVisualization from "metabase/query_builder/components/QueryVisualization";
+import ChartTypeSidebar from "metabase/query_builder/components/view/sidebars/ChartTypeSidebar";
+import { getMetadata } from "metabase/selectors/metadata";
+import { CardApi } from "metabase/services";
+import { Box, Group, Text } from "metabase/ui";
+import { PublicMode } from "metabase/visualizations/click-actions/modes/PublicMode";
+import Question from "metabase-lib/v1/Question";
+import type { Card, CardId, Dataset } from "metabase-types/api";
+
+import { useEmbeddingContext } from "../../context";
+
+interface QueryVisualizationProps {
+  questionId: CardId;
+  showVisualizationSelector?: boolean;
+}
+
+type State = {
+  loading: boolean;
+  card: Card | null;
+  cardError?: Card | string | null;
+  result: Dataset | null;
+  resultError?: Dataset | string | null;
+};
+
+export const QueryVisualizationSdk = ({
+  questionId,
+  showVisualizationSelector,
+}: QueryVisualizationProps): JSX.Element | null => {
+  const { isInitialized, isLoggedIn } = useEmbeddingContext();
+  const metadata = useSelector(getMetadata);
+
+  const [{ loading, card, result, cardError, resultError }, setState] =
+    useState<State>({
+      loading: false,
+      card: null,
+      cardError: null,
+      result: null,
+      resultError: null,
+    });
+
+  const loadCardData = async ({ questionId }: { questionId: number }) => {
+    setState(prevState => ({
+      ...prevState,
+      loading: true,
+    }));
+
+    Promise.all([
+      CardApi.get({ cardId: questionId }),
+      CardApi.query({
+        cardId: questionId,
+      }),
+    ])
+      .then(([card, result]) => {
+        setState(prevState => ({
+          ...prevState,
+          card,
+          result,
+          loading: false,
+          cardError: null,
+          resultError: null,
+        }));
+      })
+      .catch(([cardError, resultError]) => {
+        setState(prevState => ({
+          ...prevState,
+          result: null,
+          card: null,
+          loading: false,
+          cardError,
+          resultError,
+        }));
+      });
+  };
+
+  useEffect(() => {
+    if (!isInitialized || !isLoggedIn) {
+      setState({
+        loading: false,
+        card: null,
+        result: null,
+        cardError: null,
+        resultError: null,
+      });
+    } else {
```

Comment:
Oh, I like the `else` part more compared to the previous solution. I was thinking something felt off, but I couldn't think of anything else at that time.